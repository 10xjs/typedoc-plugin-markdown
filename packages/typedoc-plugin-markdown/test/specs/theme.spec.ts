import * as fs from 'fs-extra';
import { UrlMapping } from 'typedoc';

import { TestApp } from '../test-app';

describe(`Theme:`, () => {
  function getExpectedUrls(urlMappings: UrlMapping[]) {
    const expectedUrls = [];
    urlMappings.forEach((urlMapping) => {
      expectedUrls.push(urlMapping.url);
      if (urlMapping.model.children) {
        urlMapping.model.children.forEach((reflection) => {
          if (!reflection.hasOwnDocument) {
            expectedUrls.push(reflection.url);
          }
        });
      }
    });
    return expectedUrls;
  }

  let testApp: TestApp;

  beforeAll(() => {
    testApp = new TestApp(['theme.ts']);
  });

  describe(`(getUrls)`, () => {
    test(`should getUrls'`, () => {
      testApp.bootstrap();
      const urlMappings = testApp.theme.getUrls(testApp.project);
      expect(getExpectedUrls(urlMappings)).toMatchSnapshot();
    });

    test(`should getUrls with readme 'none'`, () => {
      testApp.bootstrap({ readme: 'none' });
      const urlMappings = testApp.theme.getUrls(testApp.project);
      expect(getExpectedUrls(urlMappings)).toMatchSnapshot();
    });
  });

  describe(`(isOutputDirectory)`, () => {
    let directoryListingSpy;

    beforeAll(() => {
      testApp.bootstrap();
      directoryListingSpy = jest.spyOn(fs, 'readdirSync');
    });

    test(`should test output directory true with all allowed files and directories`, () => {
      directoryListingSpy.mockReturnValue([
        '.DS_Store',
        'README.md',
        'globals.md',
        'classes',
        'enums',
        'interfaces',
        'media',
        'modules',
      ]);
      expect(testApp.theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory true with some files directories`, () => {
      directoryListingSpy.mockReturnValue([
        'README.md',
        'classes',
        'media',
        'modules',
      ]);
      expect(testApp.theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory true with just index`, () => {
      directoryListingSpy.mockReturnValue(['README.md']);
      expect(testApp.theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory false with unkown index`, () => {
      directoryListingSpy.mockReturnValue([
        'Unrecognised.md',
        'classes',
        'enums',
        'interfaces',
        'media',
        'modules',
      ]);
      expect(testApp.theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false with hidden files`, () => {
      directoryListingSpy.mockReturnValue([
        '.git',
        'classes',
        'enums',
        'interfaces',
        'media',
        'modules',
      ]);
      expect(testApp.theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false with unknown folder`, () => {
      directoryListingSpy.mockReturnValue(['README.md', 'folder']);
      expect(testApp.theme.isOutputDirectory('/path')).toBeFalsy();
    });
  });

  describe(`output directory`, () => {
    let directoryListingSpy;
    beforeAll(() => {
      directoryListingSpy = jest.spyOn(fs, 'readdirSync');
    });

    test(`should test output directory true with all allowed files and directories`, () => {
      directoryListingSpy.mockReturnValue([
        '.DS_Store',
        'README.md',
        'globals.md',
        'classes',
        'enums',
        'interfaces',
        'media',
        'modules',
      ]);
      expect(testApp.theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory true with some files directories`, () => {
      directoryListingSpy.mockReturnValue([
        'README.md',
        'classes',
        'media',
        'modules',
      ]);
      expect(testApp.theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory true with just index`, () => {
      directoryListingSpy.mockReturnValue(['README.md']);
      expect(testApp.theme.isOutputDirectory('/path')).toBeTruthy();
    });

    test(`should test output directory false with unkown index`, () => {
      directoryListingSpy.mockReturnValue([
        'Unrecognised.md',
        'classes',
        'enums',
        'interfaces',
        'media',
        'modules',
      ]);
      expect(testApp.theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false with hidden files`, () => {
      directoryListingSpy.mockReturnValue([
        '.git',
        'classes',
        'enums',
        'interfaces',
        'media',
        'modules',
      ]);
      expect(testApp.theme.isOutputDirectory('/path')).toBeFalsy();
    });

    test(`should test output directory false with unknown folder`, () => {
      directoryListingSpy.mockReturnValue(['README.md', 'folder']);
      expect(testApp.theme.isOutputDirectory('/path')).toBeFalsy();
    });
  });
});
