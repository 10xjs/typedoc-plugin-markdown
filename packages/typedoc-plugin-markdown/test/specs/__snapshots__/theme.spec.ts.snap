// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme: (getUrls) should getUrls with 'allReflectionsHaveOwnDocument' set 1`] = `
Array [
  "globals.md",
  "README.md",
  "modules/_theme_.md",
  "modules/_theme_.namespacea.md",
  "modules/_theme_.namespaceb.md",
  "enums/_theme_.enumitema.md",
  "enums/_theme_.enumitemb.md",
  "classes/_theme_.classitema.md",
  "classes/_theme_.classitemb.md",
  "interfaces/_theme_.interfaceitema.md",
  "interfaces/_theme_.interfaceitema.md#prop",
  "interfaces/_theme_.interfaceitemb.md",
  "interfaces/_theme_.interfaceitemb.md#prop",
  "variables/_theme_.declarationitema.md",
  "variables/_theme_.declarationitemb.md",
  "functions/_theme_.functionitema.md",
  "functions/_theme_.functionitemb.md",
]
`;

exports[`Theme: (getUrls) should getUrls with readme 'none' 1`] = `
Array [
  "README.md",
  "modules/_theme_.md",
  "modules/_theme_.md#declarationitema",
  "modules/_theme_.md#declarationitemb",
  "modules/_theme_.md#functionitema",
  "modules/_theme_.md#functionitemb",
  "modules/_theme_.namespacea.md",
  "modules/_theme_.namespaceb.md",
  "enums/_theme_.enumitema.md",
  "enums/_theme_.enumitemb.md",
  "classes/_theme_.classitema.md",
  "classes/_theme_.classitemb.md",
  "interfaces/_theme_.interfaceitema.md",
  "interfaces/_theme_.interfaceitema.md#prop",
  "interfaces/_theme_.interfaceitemb.md",
  "interfaces/_theme_.interfaceitemb.md#prop",
]
`;

exports[`Theme: (getUrls) should getUrls' 1`] = `
Array [
  "globals.md",
  "README.md",
  "modules/_theme_.md",
  "modules/_theme_.md#declarationitema",
  "modules/_theme_.md#declarationitemb",
  "modules/_theme_.md#functionitema",
  "modules/_theme_.md#functionitemb",
  "modules/_theme_.namespacea.md",
  "modules/_theme_.namespaceb.md",
  "enums/_theme_.enumitema.md",
  "enums/_theme_.enumitemb.md",
  "classes/_theme_.classitema.md",
  "classes/_theme_.classitemb.md",
  "interfaces/_theme_.interfaceitema.md",
  "interfaces/_theme_.interfaceitema.md#prop",
  "interfaces/_theme_.interfaceitemb.md",
  "interfaces/_theme_.interfaceitemb.md#prop",
]
`;
